File: mac/scripts/make_icon.py
#!/usr/bin/env python3
"""Generate a polished MOSS TTS app icon at every required macOS size."""

import math
import os
import sys
from pathlib import Path

from PIL import Image, ImageDraw, ImageFilter, ImageFont

SIZES = [16, 32, 64, 128, 256, 512, 1024]


def lerp(a, b, t):
    return a + (b - a) * t


def lerp_color(c1, c2, t):
    return tuple(int(lerp(c1[i], c2[i], t)) for i in range(len(c1)))


def make_icon(size: int) -> Image.Image:
    img = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    draw = ImageDraw.Draw(img)
    s = size

    # ── Background rounded rect ──────────────────────────────────────────────
    margin = s * 0.04
    r = s * 0.22   # corner radius

    # Gradient: top-left = dark navy, bottom-right = deep slate
    bg = Image.new("RGBA", (s, s), (0, 0, 0, 0))
    bg_draw = ImageDraw.Draw(bg)
    top_col    = (15, 20, 42)    # deep midnight blue
    bottom_col = (28, 15, 52)    # deep violet

    for y in range(s):
        t = y / s
        row_col = lerp_color(top_col, bottom_col, t)
        bg_draw.line([(0, y), (s - 1, y)], fill=(*row_col, 255))

    # Mask to rounded rect
    mask = Image.new("L", (s, s), 0)
    mask_draw = ImageDraw.Draw(mask)
    mask_draw.rounded_rectangle([margin, margin, s - margin, s - margin],
                                  radius=r, fill=255)
    img.paste(bg, (0, 0), mask)

    # ── Subtle inner highlight (top rim) ────────────────────────────────────
    glow = Image.new("RGBA", (s, s), (0, 0, 0, 0))
    gd = ImageDraw.Draw(glow)
    gd.rounded_rectangle(
        [margin, margin, s - margin, s * 0.55],
        radius=r,
        fill=(80, 100, 200, 18),
    )
    img = Image.alpha_composite(img, glow)

    # ── Waveform bars ────────────────────────────────────────────────────────
    # 7 bars in a balanced "mountain" profile
    bar_heights = [0.18, 0.35, 0.55, 0.78, 1.0, 0.68, 0.38, 0.22]
    n = len(bar_heights)

    total_bar_w = s * 0.62
    spacing     = total_bar_w / (n * 2 - 1)
    bar_w       = spacing * 1.0
    x_start     = (s - total_bar_w) / 2

    max_bar_h   = s * 0.46
    center_y    = s * 0.52

    for i, h_frac in enumerate(bar_heights):
        bh   = max_bar_h * h_frac
        bx   = x_start + i * (bar_w + spacing)
        by1  = center_y - bh
        by2  = center_y + bh
        br   = bar_w * 0.45    # bar corner radius

        # Per-bar gradient: cool blue → electric cyan
        t_horiz = i / (n - 1)
        top_bar  = lerp_color((50, 130, 255), (0, 220, 255), t_horiz)
        bot_bar  = lerp_color((30,  80, 200), (0, 160, 220), t_horiz)

        # Draw gradient bar via scanline
        bar_img  = Image.new("RGBA", (s, s), (0, 0, 0, 0))
        bar_draw = ImageDraw.Draw(bar_img)

        # Fill the bar shape with rounded caps
        bar_draw.rounded_rectangle(
            [bx, by1, bx + bar_w, by2],
            radius=br,
            fill=(*top_bar, 255),
        )
        # Overlay a gradient tint (lighter top, darker bottom)
        for row_y in range(int(by1), int(by2) + 1):
            tr = (row_y - by1) / max(by2 - by1, 1)
            row_col = lerp_color(top_bar, bot_bar, tr)
            bar_draw.line([(bx, row_y), (bx + bar_w, row_y)],
                          fill=(*row_col, 255))

        # Re-mask to rounded rect (gradient overwrote the mask)
        bar_mask = Image.new("L", (s, s), 0)
        bm_draw = ImageDraw.Draw(bar_mask)
        bm_draw.rounded_rectangle(
            [bx, by1, bx + bar_w, by2],
            radius=br,
            fill=255,
        )
        img = Image.alpha_composite(img, Image.composite(bar_img, Image.new("RGBA", (s, s), (0,0,0,0)), bar_mask))

    # ── Bar glow (bloom) ─────────────────────────────────────────────────────
    if size >= 64:
        bloom_layer = Image.new("RGBA", (s, s), (0, 0, 0, 0))
        bloom_draw  = ImageDraw.Draw(bloom_layer)
        for i, h_frac in enumerate(bar_heights):
            bh  = max_bar_h * h_frac
            bx  = x_start + i * (bar_w + spacing)
            t_h = i / (n - 1)
            gc  = lerp_color((80, 160, 255), (0, 230, 255), t_h)
            bloom_draw.rounded_rectangle(
                [bx - bar_w * 0.3, center_y - bh - bar_w * 0.3,
                 bx + bar_w * 1.3, center_y + bh + bar_w * 0.3],
                radius=bar_w,
                fill=(*gc, 35),
            )
        bloom_blur = bloom_layer.filter(ImageFilter.GaussianBlur(radius=max(2, size // 40)))
        img = Image.alpha_composite(img, bloom_blur)

    # ── Outer border highlight ────────────────────────────────────────────────
    border = Image.new("RGBA", (s, s), (0, 0, 0, 0))
    bd     = ImageDraw.Draw(border)
    bw     = max(1, s // 180)
    bd.rounded_rectangle(
        [margin + bw, margin + bw, s - margin - bw, s - margin - bw],
        radius=r - bw,
        outline=(120, 160, 255, 45),
        width=bw,
    )
    img = Image.alpha_composite(img, border)

    return img


def main():
    out_dir = Path(sys.argv[1]) if len(sys.argv) > 1 else Path("MossTTS.app/Contents/Resources/AppIcon.iconset")
    out_dir.mkdir(parents=True, exist_ok=True)

    for size in SIZES:
        icon = make_icon(size)
        # 1× slot
        if size <= 512:
            icon.save(out_dir / f"icon_{size}x{size}.png")
        # 2× slot (this size is the @2x for the half-size)
        if size >= 32:
            half = size // 2
            icon.save(out_dir / f"icon_{half}x{half}@2x.png")

    print(f"Icon PNGs written to {out_dir}  ({len(list(out_dir.glob('*.png')))} files)")


if __name__ == "__main__":
    main()

File: Mosslanding-Sync/Mosslanding-Sync/scripts/make_icon.swift
#!/usr/bin/env swift
// Generate a polished MOSSlanding app icon at every required macOS size.
// Native CoreGraphics port of the previous Pillow-based make_icon.py — no
// third-party deps; runs against the macOS Swift toolchain that already
// ships with Xcode CLT.
//
// Usage: swift scripts/make_icon.swift <output-iconset-dir>

import AppKit
import CoreGraphics
import CoreImage
import Foundation
import ImageIO
import UniformTypeIdentifiers

let sizes: [Int] = [16, 32, 64, 128, 256, 512, 1024]

let outDir: String = CommandLine.arguments.count > 1
    ? CommandLine.arguments[1]
    : "MOSSlanding.app/Contents/Resources/AppIcon.iconset"

@inline(__always) func lerp(_ a: Double, _ b: Double, _ t: Double) -> Double { a + (b - a) * t }

func rgb(_ r: Int, _ g: Int, _ b: Int, alpha: Double = 1.0) -> CGColor {
    CGColor(red: Double(r)/255, green: Double(g)/255, blue: Double(b)/255, alpha: alpha)
}

func makeContext(size: Int) -> CGContext {
    return CGContext(
        data: nil,
        width: size, height: size,
        bitsPerComponent: 8, bytesPerRow: 0,
        space: CGColorSpaceCreateDeviceRGB(),
        bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
    )!
}

func makeIcon(size: Int) -> CGImage {
    let ctx = makeContext(size: size)
    let s = CGFloat(size)

    // Flip y so coordinates read top-down (matches the Pillow version).
    ctx.translateBy(x: 0, y: s)
    ctx.scaleBy(x: 1, y: -1)

    let margin = s * 0.04
    let cornerR = s * 0.22
    let bgRect = CGRect(x: margin, y: margin, width: s - 2 * margin, height: s - 2 * margin)
    let bgPath = CGPath(roundedRect: bgRect, cornerWidth: cornerR, cornerHeight: cornerR, transform: nil)

    // ── Background gradient (midnight blue → deep violet) ─────────────────
    ctx.saveGState()
    ctx.addPath(bgPath)
    ctx.clip()
    let bgGradient = CGGradient(
        colorsSpace: CGColorSpaceCreateDeviceRGB(),
        colors: [rgb(15, 20, 42), rgb(28, 15, 52)] as CFArray,
        locations: [0.0, 1.0]
    )!
    ctx.drawLinearGradient(
        bgGradient,
        start: CGPoint(x: 0, y: 0),
        end:   CGPoint(x: 0, y: s),
        options: []
    )

    // Subtle inner top-rim highlight.
    let topRim = CGRect(x: margin, y: margin, width: s - 2 * margin, height: s * 0.55 - margin)
    let topRimPath = CGPath(roundedRect: topRim, cornerWidth: cornerR, cornerHeight: cornerR, transform: nil)
    ctx.addPath(topRimPath)
    ctx.setFillColor(rgb(80, 100, 200, alpha: 0.07))
    ctx.fillPath()
    ctx.restoreGState()

    // ── Waveform bars ─────────────────────────────────────────────────────
    let barHeights: [CGFloat] = [0.18, 0.35, 0.55, 0.78, 1.0, 0.68, 0.38, 0.22]
    let n = barHeights.count
    let totalBarW = s * 0.62
    let spacing = totalBarW / (CGFloat(n) * 2 - 1)
    let barW = spacing
    let xStart = (s - totalBarW) / 2
    let maxBarH = s * 0.46
    let centerY = s * 0.52

    for (i, hFrac) in barHeights.enumerated() {
        let bh = maxBarH * hFrac
        let bx = xStart + CGFloat(i) * (barW + spacing)
        let by1 = centerY - bh
        let by2 = centerY + bh
        let br = barW * 0.45

        let tH = Double(i) / Double(n - 1)
        let topColor = CGColor(
            red:   lerp( 50/255, 0/255,   tH),
            green: lerp(130/255, 220/255, tH),
            blue:  lerp(255/255, 255/255, tH),
            alpha: 1
        )
        let botColor = CGColor(
            red:   lerp( 30/255, 0/255,   tH),
            green: lerp( 80/255, 160/255, tH),
            blue:  lerp(200/255, 220/255, tH),
            alpha: 1
        )
        let barGradient = CGGradient(
            colorsSpace: CGColorSpaceCreateDeviceRGB(),
            colors: [topColor, botColor] as CFArray,
            locations: [0.0, 1.0]
        )!

        let barRect = CGRect(x: bx, y: by1, width: barW, height: by2 - by1)
        let barPath = CGPath(roundedRect: barRect, cornerWidth: br, cornerHeight: br, transform: nil)

        ctx.saveGState()
        ctx.addPath(barPath)
        ctx.clip()
        ctx.drawLinearGradient(
            barGradient,
            start: CGPoint(x: 0, y: by1),
            end:   CGPoint(x: 0, y: by2),
            options: []
        )
        ctx.restoreGState()
    }

    // ── Bar bloom (Gaussian) — only on icons big enough to show it ───────
    if size >= 64 {
        let bloomCtx = makeContext(size: size)
        bloomCtx.translateBy(x: 0, y: s)
        bloomCtx.scaleBy(x: 1, y: -1)
        for (i, hFrac) in barHeights.enumerated() {
            let bh = maxBarH * hFrac
            let bx = xStart + CGFloat(i) * (barW + spacing)
            let tH = Double(i) / Double(n - 1)
            let color = CGColor(
                red:   lerp( 80/255,  0/255,   tH),
                green: lerp(160/255, 230/255, tH),
                blue:  lerp(255/255, 255/255, tH),
                alpha: 0.14
            )
            bloomCtx.setFillColor(color)
            let bloomRect = CGRect(
                x: bx - barW * 0.3,
                y: centerY - bh - barW * 0.3,
                width: barW * 1.6,
                height: 2 * bh + barW * 0.6
            )
            let bloomPath = CGPath(roundedRect: bloomRect, cornerWidth: barW, cornerHeight: barW, transform: nil)
            bloomCtx.addPath(bloomPath)
            bloomCtx.fillPath()
        }
        let bloomImage = bloomCtx.makeImage()!
        let radius = max(2.0, Double(size) / 40.0)
        let ci = CIContext()
        let blurred = CIImage(cgImage: bloomImage)
            .clampedToExtent()
            .applyingFilter("CIGaussianBlur", parameters: ["inputRadius": radius])
            .cropped(to: CGRect(x: 0, y: 0, width: size, height: size))
        if let blurredCG = ci.createCGImage(blurred, from: blurred.extent) {
            // Undo the flip so the upright bloom composites correctly.
            ctx.saveGState()
            ctx.scaleBy(x: 1, y: -1)
            ctx.translateBy(x: 0, y: -s)
            ctx.draw(blurredCG, in: CGRect(x: 0, y: 0, width: s, height: s))
            ctx.restoreGState()
        }
    }

    // ── Outer rim ─────────────────────────────────────────────────────────
    let bw = max(1.0, Double(size) / 180.0)
    let rimRect = bgRect.insetBy(dx: CGFloat(bw), dy: CGFloat(bw))
    let rimPath = CGPath(
        roundedRect: rimRect,
        cornerWidth: cornerR - CGFloat(bw),
        cornerHeight: cornerR - CGFloat(bw),
        transform: nil
    )
    ctx.addPath(rimPath)
    ctx.setStrokeColor(rgb(120, 160, 255, alpha: 0.18))
    ctx.setLineWidth(CGFloat(bw))
    ctx.strokePath()

    return ctx.makeImage()!
}

func savePNG(_ image: CGImage, to path: String) {
    let url = URL(fileURLWithPath: path)
    let type: CFString = UTType.png.identifier as CFString
    guard let dest = CGImageDestinationCreateWithURL(url as CFURL, type, 1, nil) else {
        FileHandle.standardError.write(Data("could not open \(path) for write\n".utf8))
        return
    }
    CGImageDestinationAddImage(dest, image, nil)
    if !CGImageDestinationFinalize(dest) {
        FileHandle.standardError.write(Data("could not finalize PNG \(path)\n".utf8))
    }
}

let fm = FileManager.default
try? fm.createDirectory(atPath: outDir, withIntermediateDirectories: true)

for size in sizes {
    let icon = makeIcon(size: size)
    if size <= 512 {
        savePNG(icon, to: "\(outDir)/icon_\(size)x\(size).png")
    }
    if size >= 32 {
        let half = size / 2
        savePNG(icon, to: "\(outDir)/icon_\(half)x\(half)@2x.png")
    }
}

let files = (try? fm.contentsOfDirectory(atPath: outDir).filter { $0.hasSuffix(".png") }) ?? []
print("Icon PNGs written to \(outDir)  (\(files.count) files)")
